### PR TITLE
include working splinelib++ classes and tests

### PR DESCRIPTION
**Pull request by cm314**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -18,4 +18,5 @@ cmake_minimum_required(VERSION 3.6 FATAL_ERROR)
 project(SplineLib VERSION 0.0.1 LANGUAGES CXX)
 set(CMAKE_CXX_STANDARD 17)
 
-add_subdirectory(test)
\ No newline at end of file
+add_subdirectory(test)
+add_subdirectory(src)
```
Comment (mfcats) [inline CMakeLists.txt]: Can you add an empty line at the end, please.

File: CMakeLists.txt [hunk 2]
```diff
@@ -18,4 +18,5 @@ cmake_minimum_required(VERSION 3.6 FATAL_ERROR)
 project(SplineLib VERSION 0.0.1 LANGUAGES CXX)
 set(CMAKE_CXX_STANDARD 17)
 
-add_subdirectory(test)
\ No newline at end of file
+add_subdirectory(test)
```
Comment (mfcats) [inline CMakeLists.txt]: Please change the order of test and src. test depends on src. Therefore, src should be in front of test.

File: src/b_spline.cc [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+#include "b_spline.h"
+
+#include <algorithm>
+#include <functional>
+#include <numeric>
+
+#include "basis_function_factory.h"
+
+BSpline::BSpline(KnotVector knot_vector, Degree degree, std::vector<ControlPoint> control_points) :
```
Comment (mfcats) [inline src/b_spline.cc]: Please pass `knot_vector` and `control_points` as const reference. This should increase performance.

File: src/b_spline_basis_function.cc [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+You should have received a copy of the GNU Lesser General Public License along with SplineLib.  If not, see
+<http://www.gnu.org/licenses/>.
+*/
+
+#include "b_spline_basis_function.h"
+
+#include "basis_function_factory.h"
+
+BSplineBasisFunction::BSplineBasisFunction(KnotVector knot_vector,
```
Comment (mfcats) [inline src/b_spline_basis_function.cc]: `knot_vector` should be passed by const reference.

File: src/b_spline_basis_function.h [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+*/
+
+#ifndef SPLINELIB_BSPLINEBASISFUNCTION_H
+#define SPLINELIB_BSPLINEBASISFUNCTION_H
+
+#include <cmath>
+#include <limits>
+#include <memory>
+
+#include <iostream>
```
Comment (mfcats) [inline src/b_spline_basis_function.h]: Is `iostream` required?

File: src/b_spline_basis_function.h [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+ protected:
+  double EvaluateOnSupport(double param_coord) const override;
+  double EvaluateDerivativeOnSupport(Derivative derivative,
+                                     double param_coord) const override;
+
+  std::unique_ptr<BasisFunction> left_lower_degree_;
+  std::unique_ptr<BasisFunction> right_lower_degree_;
+
+ private:
+  void SetLowerDegreeBsisFunctions(KnotVector knot_vector,
```
Comment (mfcats) [inline src/b_spline_basis_function.h]: It should be `SetLowerDegreeBasisFunctions`.

File: src/basis_function.cc [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+double BasisFunction::EvaluateDerivative(Derivative derivative,
+                                         double param_coord) const {
+  return derivative == 0 ? Evaluate(param_coord) : IsCoordinateInSupport(param_coord)
+                                                   ? this->EvaluateDerivativeOnSupport(derivative,
+                                                                                       param_coord)
+                                                   : 0.0;
+}
+
+BasisFunction::BasisFunction(KnotVector knot_vector,
```
Comment (mfcats) [inline src/basis_function.cc]: `knot_vector` should be passed by const reference.

File: src/basis_function_factory.cc [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+#include "b_spline_basis_function.h"
+#include "zero_degree_b_spline_basis_function.h"
+
+BasisFunction *BasisFunctionFactory::CreateDynamic(KnotVector knot_vector,
+                                                   uint64_t start_of_support,
+                                                   Degree degree) const {
+  if (degree < 0) throw std::runtime_error("Basis function degree must be positive.");
+  if (degree == 0) return new ZeroDegreeBSplineBasisFunction(knot_vector, start_of_support);
+  return new BSplineBasisFunction(knot_vector, degree, start_of_support);
+}
```
Comment (mfcats) [inline src/basis_function_factory.cc]: Please add newline at the end of file.

File: src/control_point.cc [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+License as published by the Free Software Foundation version 3 of the License.
+
+SplineLib is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty
+of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along with SplineLib.  If not, see
+<http://www.gnu.org/licenses/>.
+*/
+
+#include <iostream>
```
Comment (mfcats) [inline src/control_point.cc]: Is `iostream` required?

File: src/control_point.cc [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+#include "control_point.h"
+
+ControlPoint::ControlPoint(std::initializer_list<double> coordinates) : coordinates_(coordinates) {}
+
+ControlPoint::ControlPoint(std::vector<double> coordinates) : coordinates_(coordinates) {}
```
Comment (mfcats) [inline src/control_point.cc]: `coordinates` should be passed by const reference.

File: src/control_point.cc [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+
+int ControlPoint::dimension() const {
+  return coordinates_.size();
```
Comment (mfcats) [inline src/control_point.cc]: This should be casted to `int`.

File: src/knot_vector.cc [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
+#include "knot_vector.h"
+
+#include <algorithm>
+#include <cmath>
+#include <functional>
+
+#include "numeric_settings.h"
+
+KnotVector::KnotVector(std::vector<double> knots) : knots_(std::move(knots)) {}
```
Comment (mfcats) [inline src/knot_vector.cc]: `knots` should be passed by const reference.

File: src/knot_vector.h [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+#ifndef SPLINELIB_KNOTVECTOR_H
+#define SPLINELIB_KNOTVECTOR_H
+
+#include <initializer_list>
+#include <limits>
+#include <stdexcept>
+#include <utility>
+#include <vector>
+
+#include <iostream>
```
Comment (mfcats) [inline src/knot_vector.h]: Is `iostream` required.

File: src/knot_vector.h [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+typedef std::vector<double>::const_iterator ConstKnotIterator;
+
+class KnotVector {
+ public:
+  KnotVector() = default;
+  explicit KnotVector(std::vector<double> knots);
+  KnotVector(std::initializer_list<double> knots);
```
Comment (mfcats) [inline src/knot_vector.h]: This constructor should have `explicit` keyword.

File: src/zero_degree_b_spline_basis_function.cc [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+
+You should have received a copy of the GNU Lesser General Public License along with SplineLib.  If not, see
+<http://www.gnu.org/licenses/>.
+*/
+
+#include <cmath>
+
+#include "numeric_settings.h"
+
+#include "zero_degree_b_spline_basis_function.h"
```
Comment (mfcats) [inline src/zero_degree_b_spline_basis_function.cc]: This should be the first included header.

File: src/zero_degree_b_spline_basis_function.cc [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+
+ZeroDegreeBSplineBasisFunction::ZeroDegreeBSplineBasisFunction(KnotVector knot_vector,
```
Comment (mfcats) [inline src/zero_degree_b_spline_basis_function.cc]: `knot_vector` should be passed by const reference.

File: test/CMakeLists.txt [hunk 17]
```diff
@@ -14,9 +14,13 @@
...
+        zero_degree_b_spline_basis_function_test.cc
+        b_spline_basis_function_test.cc
+        b_spline_test.cc)
 
 find_package(GTest REQUIRED)
 
 add_executable(SplineLibTests ${TEST_SOURCES})
-target_link_libraries(SplineLibTests GTest::Main)
\ No newline at end of file
+target_link_libraries(SplineLibTests SplineLib GTest::Main)
```
Comment (mfcats) [inline test/CMakeLists.txt]: Please add an empty line at the end.

File: src/knot_vector.h (hunk 14 above)
Comment (cm314) [reply src/knot_vector.h]: This leads to the error message:
SplineLib/test/b_spline_test.cc(25): error: copy-list-initialization cannot use a constructor marked "explicit"
      KnotVector knot_vector = {0, 0, 0, 1, 2, 3, 4, 4, 5, 5, 5};

**Pull request merged**